### PR TITLE
Add background to breadcrumbs and status bar

### PR DESCRIPTION
**Pull request by s9ferech**
### Pull Request Description

This adds a background to the breadcrumbs and statusbar, solving issue #1398. Background color and corner radius for each can be set through the theme.

<kbd></kbd>

### Important Notes

The implementation is strongly coupled with the internals of breadcrumbs and the status bar. This is necessary to compute the width and height of those components and to track their changes. We could improve this in the future through introduction of a generic layout mechanism.

### Checklist
Please include the following checklist in your PR:

- [x] The `CHANGELOG.md` was updated with the changes introduced in this PR.
- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/enso-org/enso/blob/main/docs/style-guide/rust.md) style guide.
- [ ] All code has automatic tests where possible.
- [ ] All code has been profiled where possible.
- [x] All code has been manually tested in the IDE.
- [x] All code has been manually tested in the "debug/interface" scene.
- [ ] All code has been manually tested by the PR owner against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).
- [ ] All code has been manually tested by at least one reviewer against our [test scenarios](https://docs.google.com/spreadsheets/d/1RatJDM_f9_3bvYhl3Bpq2d8SyKgtVdrV1RkGxPU17c8/edit?ts=5faa7049#gid=0).

### CODE REVIEW

File: src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs [hunk 1]
```diff
@@ -10,30 +10,47 @@ pub use breadcrumb::Breadcrumb;
...
+                                             + breadcrumb::PADDING
+                                             + breadcrumb::VERTICAL_MARGIN
+                                             + VERTICAL_MARGIN;
+
+const MACOS_TRAFFIC_LIGHTS_CONTENT     : f32 = 52.0;
+const MACOS_TRAFFIC_LIGHTS_SIDE_OFFSET : f32 = 13.0;
+const MACOS_TRAFFIC_LIGHTS_WIDTH       : f32 =
+    MACOS_TRAFFIC_LIGHTS_CONTENT + 2.0 * MACOS_TRAFFIC_LIGHTS_SIDE_OFFSET;
+// This should be as large as the shadow around the background.
+const MAGIC_SHADOW_MARGIN              : f32 = 40.0;
```
Comment (s9ferech) [inline src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs]: I use this constant to estimate how much space the drop shadow will need. The implementations of list view, nodes and tooltip seem to do something similar. It would be good if we could get this from the definition of the shadow or if we didn't need to mention this value at all. @farmaazon do you have a suggestion what is the best way to solve this?

File: src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs [hunk 2]
```diff
@@ -10,30 +10,47 @@ pub use breadcrumb::Breadcrumb;
...
+const VERTICAL_MARGIN                  : f32 = GLYPH_WIDTH;
+const HORIZONTAL_MARGIN                : f32 = GLYPH_WIDTH;
+const TEXT_SIZE                        : f32 = 12.0;
+const HEIGHT                           : f32 = VERTICAL_MARGIN
+                                             + breadcrumb::VERTICAL_MARGIN
+                                             + breadcrumb::PADDING
+                                             + LINE_HEIGHT
+                                             + breadcrumb::PADDING
+                                             + breadcrumb::VERTICAL_MARGIN
+                                             + VERTICAL_MARGIN;
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs]: we are aligning lines only if there are glued together. In such a scenario it sees very strange, so I'd skip aligning this top block to bottom one

Comment (wdanilo) [reply src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs]: Btw, I love how you have written it - keeping the visual hierarchy of what goes after what. Very clear.

File: src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs [hunk 3]
```diff
@@ -181,15 +223,33 @@ impl BreadcrumbsModel {
...
+    fn breadcrumbs_container_size(&self) -> f32 {
         self.breadcrumbs.borrow().iter().map(|breadcrumb| breadcrumb.width()).sum()
     }
 
-    fn relayout_for_project_name_width(&self, width:f32) {
-        self.breadcrumbs_container.set_position_x(width.round());
+    fn update_layout(&self) {
+        let is_macos     = ARGS.platform.map(|p|p.is_macos()) == Some(true);
+        let is_frameless = ARGS.frame == Some(false);
+        let x_offset     = if is_macos && is_frameless { MACOS_TRAFFIC_LIGHTS_WIDTH }
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs]: Why did you move this logic into breadcrumbs? I don't think this component is responsible for making the spacing on the left side. We will use breadcrumbs in the future in multiple panels, not only on top of the window. So to correctly position breadcrumbs is the responsibility of its parent, not this component itself

**Review by wdanilo — COMMENTED** [review]
Just minor code comments for now. I'd love to see it again after the MacOS traffic lights problem gets resolved that I was writing about in the comment. In case you feel that the proposed design is bad, please reach out to me and lets talk about it! :) 

Also, from visual perspective:
1. The backgrounds were meant to not go outside of the window, they should be contained inside (the shadows can go outside, but the bqckgrounds should be fully visible). There should be a padding from left side of window border and bottom side of window border to the background of the status bar. The same applies to breadcrumbs. This way:
  image
2. Make the shadow of these panels controllable from the theme editor as well. We might want to adjust them slightly. Make in themes `status_bar.background.shadow` and all subsequent parameters (size, spread, fading, etc) and set them just to the generic ones, like:
```
    background {
        shadow = shadow, shadow;
        shadow {
            size = shadow::size, shadow::size;
            ...
        }
    }
...
```

File: src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs (hunk 3 above)
Comment (s9ferech) [reply src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs]: Because in my implementation, the breadcrumb object stretched to contain the traffic lights. Also, I found the positioning a bit weird as it was before. The breadcrumbs seemed to position themselves in the corner of the screen, but we added this offset from the outside. It just isn't what I would expect.

File: src/rust/ide/view/graph-editor/src/lib.rs [hunk 4]
```diff
@@ -1273,9 +1273,8 @@ impl GraphEditorModel {
         let is_macos     = ARGS.platform.map(|p|p.is_macos()) == Some(true);
         let is_frameless = ARGS.frame == Some(false);
         let x_offset     = if is_macos && is_frameless { MACOS_TRAFFIC_LIGHTS_WIDTH }
-                           else                        { MACOS_TRAFFIC_LIGHTS_SIDE_OFFSET };
+                           else                        { 0.0 };
```
Comment (s9ferech) [inline src/rust/ide/view/graph-editor/src/lib.rs]: I changed this because I don't see why we should have the offset if the traffic lights are not their. Please correct me if there is a reason for the offset. Also, someone with Mac OS should check if the new design still works there.

File: src/rust/ensogl/lib/components/src/shadow.rs [hunk 5]
```diff
@@ -2,47 +2,79 @@
 use crate::prelude::*;
 
 use ensogl_core::data::color;
-use ensogl_core::display::DomSymbol;
+use ensogl_core::display::{DomSymbol, style};
```
Comment (wdanilo) [inline src/rust/ensogl/lib/components/src/shadow.rs]: we're not using such imports (Style guide)

File: src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs [hunk 6]
```diff
@@ -57,10 +69,27 @@ mod background {
...
-            let gray     = 34.0/255.0;
-            let bg_color = color::Rgba::new(gray,gray,gray,1.0);
-            Plane().fill(bg_color).into()
+        (style:Style) {
+            let width             = Var::<Pixels>::from("input_size.x");
+            let height            = Var::<Pixels>::from("input_size.y");
+
+            let corner_radius     = style.get_number
+                                    (ensogl_theme::graph_editor::breadcrumbs::background
+                                    ::corner_radius);
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs]: extract `ensogl_theme::graph_editor::breadcrumbs` as `breadcrumbsTheme` and make these nulti-line expressions single line

File: src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs [hunk 7]
```diff
@@ -57,10 +69,27 @@ mod background {
...
+            let shape_width       = width - MAGIC_SHADOW_MARGIN.px() * 2.0;
+            let shape_height      = height - MAGIC_SHADOW_MARGIN.px() * 2.0;
+            let shape             = Rect((&shape_width,&shape_height))
+                                    .corners_radius(corner_radius.px());
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs]: extract variables if possible to have one line expressions here and in other places. Its easier to read them if the components are well named variables

File: src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs [hunk 8]
```diff
@@ -138,7 +167,7 @@ pub struct BreadcrumbsModel {
     display_object        : display::object::Instance,
     background            : background::View,
     project_name          : ProjectName,
-    root                   : display::object::Instance,
+    root: display::object::Instance,
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs]: why this change?

File: src/rust/ide/view/graph-editor/src/lib.rs (hunk 4 above)
Comment (wdanilo) [reply src/rust/ide/view/graph-editor/src/lib.rs]: This was added here because we want to have the same padding of breadcrumbs fron the left side of the window as the padding between left side of the window and traffic lights (in fact we want to use the same padding for top padding too). Please adjust it so these paddings are the same.

File: src/rust/ide/view/src/status_bar.rs [hunk 9]
```diff
@@ -75,6 +83,40 @@ pub mod process {
...
+
+            let corner_radius     = style.get_number
+                                    (ensogl_theme::application::status_bar::background::corner_radius);
+            let shape_width       = width  - MAGIC_SHADOW_MARGIN.px() * 2.0;
+            let shape_height      = height - MAGIC_SHADOW_MARGIN.px() * 2.0;
+            let shape             = Rect((&shape_width,&shape_height))
+                                    .corners_radius(corner_radius.px());
+
+            let bg_color          = style.get_color
+                                    (ensogl_theme::application::status_bar::background);
```
Comment (wdanilo) [inline src/rust/ide/view/src/status_bar.rs]: make all these utliline expressions single lines please

File: src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs (hunk 8 above)
Comment (s9ferech) [reply src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs]: This was just a refactoring accident.

File: src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs [hunk 10]
```diff
@@ -181,15 +214,27 @@ impl BreadcrumbsModel {
...
+        self.breadcrumbs_container.set_position_x(
+            HORIZONTAL_PADDING + self.project_name.width.value().round());
+
+        let background_width  = HORIZONTAL_PADDING
+                              + self.project_name.width.value().round()
+                              + self.breadcrumbs_container_size()
+                              + HORIZONTAL_PADDING;
+        let background_height = HEIGHT;
+        self.background.size.set(Vector2(background_width+MAGIC_SHADOW_MARGIN*2.0,
+            background_height+MAGIC_SHADOW_MARGIN*2.0));
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs]: pls refactor it to var and make this one-liners

File: src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs [hunk 11]
```diff
@@ -181,15 +214,27 @@ impl BreadcrumbsModel {
...
+    fn breadcrumbs_container_size(&self) -> f32 {
         self.breadcrumbs.borrow().iter().map(|breadcrumb| breadcrumb.width()).sum()
     }
 
-    fn relayout_for_project_name_width(&self, width:f32) {
-        self.breadcrumbs_container.set_position_x(width.round());
+    fn update_layout(&self) {
+        self.project_name.set_position_x(HORIZONTAL_PADDING);
+        self.breadcrumbs_container.set_position_x(
+            HORIZONTAL_PADDING + self.project_name.width.value().round());
```
Comment (wdanilo) [inline src/rust/ide/view/graph-editor/src/component/breadcrumbs.rs]: pls refactor it to var and make this one-liners

**Review by wdanilo — APPROVED** [review]
